cmd/asset-syncer/utils.go

```diff
@@ -431,77 +445,127 @@ func extractFilesFromBuffer(buf *bytes.Buffer) (*artifactFiles, error) {
+		Icon:          chartMetadata.Icon,
+		Category:      chartMetadata.Annotations["category"],
+		ChartVersions: []models.ChartVersion{chartVersion},
+	}, nil
+}
+
+func chartImportWorker(repoURL *url.URL, r *OCIRegistry, chartJobs <-chan pullChartJob, resultChan chan pullChartResult) {
+	for j := range chartJobs {
+		log.WithFields(log.Fields{"name": j.AppName, "tag": j.Tag}).Debug("pulling chart")
+		chart, err := pullAndExtract(repoURL, j.AppName, j.Tag, r.puller, r)
+		resultChan <- pullChartResult{chart, err}
+	}
+}
+
 // Charts retrieve the list of charts exposed in the repo
 func (r *OCIRegistry) Charts() ([]models.Chart, error) {
-	result := []models.Chart{}
+	result := map[string]models.Chart{}
 	url, err := parseRepoURL(r.RepoInternal.URL)
 	if err != nil {
 		return nil, err
 	}
 
+	chartJobs := make(chan pullChartJob, numWorkers)
+	chartChan := make(chan pullChartResult)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	totalCharts := 0
 	for _, appName := range r.repositories {
-		chartInfo := models.Chart{}
-		firstTag := true
-		for _, tag := range r.tags[appName].Tags {
-			// Pull Chart
-			// TODO: Run in parallel
-			ref := path.Join(url.Host, url.Path, fmt.Sprintf("%s:%s", appName, tag))
-			chartBuffer, digest, err := r.puller.PullOCIChart(ref)
-			if err != nil {
-				return nil, err
-			}
-
-			// Extract
-			files, err := extractFilesFromBuffer(chartBuffer)
-			if err != nil {
-				return nil, err
-			}
-			chartMetadata := h3chart.Metadata{}
-			err = yaml.Unmarshal([]byte(files.Metadata), &chartMetadata)
-			if err != nil {
-				return nil, err
-			}
-
-			// Format Data
-			chartVersion := models.ChartVersion{
-				Version:    chartMetadata.Version,
-				AppVersion: chartMetadata.AppVersion,
-				Digest:     digest,
-				URLs:       chartMetadata.Sources,
-				Readme:     files.Readme,
-				Values:     files.Values,
-				Schema:     files.Schema,
-			}
-			if firstTag {
-				firstTag = false
-				maintainers := []chart.Maintainer{}
-				for _, m := range chartMetadata.Maintainers {
-					maintainers = append(maintainers, chart.Maintainer{
-						Name:  m.Name,
-						Email: m.Email,
-						Url:   m.URL,
-					})
-				}
-				chartInfo = models.Chart{
-					ID:            path.Join(r.Name, appName),
-					Name:          appName,
-					Repo:          &models.Repo{Namespace: r.Namespace, Name: r.Name, URL: r.URL, Type: r.Type},
-					Description:   chartMetadata.Description,
-					Home:          chartMetadata.Home,
-					Keywords:      chartMetadata.Keywords,
-					Maintainers:   maintainers,
-					Sources:       chartMetadata.Sources,
-					Icon:          chartMetadata.Icon,
-					Category:      chartMetadata.Annotations["category"],
-					ChartVersions: []models.ChartVersion{chartVersion},
+		totalCharts += len(r.tags[appName].Tags)
+	}
+	// Process 10 charts at a time
+	for i := 0; i < numWorkers; i++ {
+		go chartImportWorker(url, r, chartJobs, chartChan)
+	}
+
+	// Start receiving charts
+	processedCharts := 0
+	go func() {
```

Comment:
This code receiving on `chartChan` now does not need to be a separate go-routine, it just needs to happen after you've started feeding the workers below (which *would* be in a goroutine, as it blocks). Now that `chartChan` is closed when all the data has been loaded the range will exit naturally when `chartChan` is exhausted.

Anyway, I'm not saying it needs to be exactly like that, but I do think we should be using the waitgroup to do the synchronisation of the workers naturally somehow, rather than setting it to 1 and counting etc.